scriptlets/clean_machine

```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+# this is a generic clean script that deletes any thing that may break the
+# current test run due to a previous run. This is especially useful for
+# noprovision machines.
+
+# We can do all these things without checking the result as we may be running
+# in any situation, so the previous run could have installed a snap, deb, source
+# etc.
+
+# First lets check that launching this was intentional
+
+if [ "$1" = "--im-sure" ]; then
+  echo "Cleaning the machine"
+else
+  echo "This is dangerous as it tries to purge an environment before running"
+  echo "the tests. You should not launch this on your own machine."
+  echo "if you are sure launch: `basename "$0"` --im-sure"
+  exit 1
+fi
+
+set +e
+
+# delete all checkbox snaps, this includes any frontend or custom frontend
+# (they all contain checkbox in their name). This should also purge any
+# running checkbox*agent service
+CHECKBOX_SNAPS=$(snap list | \grep -E 'checkbox' | \grep -P -o "^[\w-]+")
+if [ -n "$CHECKBOX_SNAPS" ]; then
+    echo "Found the following Checkbox snaps"
+    echo "  " $CHECKBOX_SNAPS
+    echo "Deleting them..."
+    snap remove $CHECKBOX_SNAPS
+fi
+
+# delete all installed Checkbox debian packages. This should also purge any
+# previously installed provider
+CHECKBOX_DEBS=$(apt list --installed | \grep checkbox | \grep -P -o "^[\w-]+")
+if [ -n "$CHECKBOX_DEBS" ]; then
+    echo "Found the following Checkbox debs"
+    echo "  " $CHECKBOX_DEBS
+    echo "Deleting them..."
+    apt purge -y -qq $CHECKBOX_DEBS
+fi
+
+# delete old sessions, as Checkbox will try to resume them when started,
+# but this is a new test session, so this is not desirable
+if [ -d "/var/tmp/checkbox-ng" ]; then
+  echo "Removing all old sessions still on the machine"
+  rm -rf /var/tmp/checkbox-ng
+fi
+
+# delete any provider that was globally "developed" (installed in editable mode)
+if [ -d "/var/tmp/checkbox-providers-develop/"]; then
+  echo "Removing all developed providers"
+  rm -rf /var/tmp/checkbox-providers-develop
+fi
+
+# remove all ppa repos pre-configured as they may make the installer install
+# from the wrong source
+# leave also /ppa and /testing, legacy names of /edge and /beta
+echo "Trying to remove all Checkbox ppas"
+add-apt-repository --remove -y ppa:checkbox-dev/ppa
+add-apt-repository --remove -y ppa:checkbox-dev/testing
+add-apt-repository --remove -y ppa:checkbox-dev/edge
+add-apt-repository --remove -y ppa:checkbox-dev/beta
+add-apt-repository --remove -y ppa:checkbox-dev/stable
+
+# Also remove any cloned version of Checkbox in any home, as this will clash
+# with anything trying to provision from source
+CHECKBOX_SOURCE_CLONES=$(find /home/ -maxdepth 2 -name "*checkbox*" -type d)
```

Comment:
Just a suggestion but you can delete the directories you find directly within `find`
```
echo "Removing all clones of Checkbox on the machine"
find /home/ -maxdepth 2 -name "*checkbox*" -type d -exec rm -rf {} +
```
If I remember correctly this should also display the ones that are found and removed.

Reply:
yes, I wanted for "simmetry" to only print the "Removing all clones..." message when there is actually something to remove, but I can change this.

Reply:
Just to clarify: you decided against using the `-exec` version?

Reply:
yes as I think the current one produces a better output log